Conan on Mac can't recognize compiler correctly

Hello, 
https://travis-ci.org/procxx/kepka/jobs/322618396

on the link you can see that Conan recognize compiler as gcc 4.2 on Travis, but it's not true.

I tried passing `-s compiler.version=7.2` to force detection override, subsequently cmake files because conan generates invalid compiler identification:

```
-- Current conanbuildinfo.cmake directory: /Users/travis/build/procxx/kepka/_build_
CMake Error at _build_/conanbuildinfo.cmake:474 (message):
  Incorrect 'gcc', is not the one detected by CMake: 'AppleClang'
Call Stack (most recent call first):
  _build_/conanbuildinfo.cmake:173 (conan_check_compiler)
  CMakeLists.txt:30 (conan_basic_setup)
```

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -156,14 +156,14 @@ def _get_default_compiler(output):
         command = cc or cxx
         if "gcc" in command:
             gcc = _gcc_compiler(output, command)
-            if gcc is not None:
-                return gcc
-            if platform.system() == "Darwin":
+            if platform.system() == "Darwin" and gcc is None:
                 output.warn(
                     "%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command
                 )
-                # using clang instead
+                # Fallback to use clang instead
                 command = "clang"
+            else:
+                return gcc
         if "clang" in command.lower():
             return _clang_compiler(output, command)
         if platform.system() == "SunOS" and command.lower() == "cc":

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -24,6 +24,14 @@ def _execute(command):
 
 
 def _gcc_compiler(output, compiler_exe="gcc"):
+
+    if platform.system() == "Darwin":
+        # In Mac OS X check if gcc is a fronted using apple-clang
+        _, out = _execute("%s --version" % compiler_exe)
+        out = out.lower()
+        if "clang" in out:
+            return None
+
     try:
         _, out = _execute('%s -dumpversion' % compiler_exe)
         compiler = "gcc"
@@ -147,16 +155,15 @@ def _get_default_compiler(output):
         output.info("CC and CXX: %s, %s " % (cc or "None", cxx or "None"))
         command = cc or cxx
         if "gcc" in command:
-            if platform.system() != "Darwin":
-                return _gcc_compiler(output, command)
-
-            # In Mac OS X check if gcc is a fronted using apple-clang
-            _, out = _execute("%s --version" % command)
-            out = out.lower()
-            if "clang" not in out:
-                return _gcc_compiler(output, command)
-            output.warn("%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command)
-            command = "clang"  # Force use of clang when gcc is detected as a fronted of apple-clang
+            gcc = _gcc_compiler(output, command)
+            if gcc is not None:
+                return gcc
+            if platform.system() == "Darwin":
+                output.warn(
+                    "%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command
+                )
+                # using clang instead
+                command = "clang"
         if "clang" in command.lower():
             return _clang_compiler(output, command)
         if platform.system() == "SunOS" and command.lower() == "cc":

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -157,13 +157,10 @@ def _get_default_compiler(output):
         if "gcc" in command:
             gcc = _gcc_compiler(output, command)
             if platform.system() == "Darwin" and gcc is None:
-                output.warn(
-                    "%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command
+                output.error(
+                    "%s detected as a frontend using apple-clang. Compiler not supported" % command
                 )
-                # Fallback to use clang instead
-                command = "clang"
-            else:
-                return gcc
+            return gcc
         if "clang" in command.lower():
             return _clang_compiler(output, command)
         if platform.system() == "SunOS" and command.lower() == "cc":

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -23,19 +23,6 @@ def _execute(command):
     return proc.returncode, "".join(output_buffer)
 
 
-def _is_real_compiler(compiler_name, compiler_exe):
-    """
-    In some platforms gcc compiler is not a real gcc, i.e: in Mac OS X gcc is clang by default.
-    It can also be a problem if gcc is a symlink to clang or the other way around.
-    """
-
-    try:
-        _, out = _execute("%s --version" % compiler_exe)
-        return compiler_name in out
-    except Exception:
-        return False
-
-
 def _gcc_compiler(output, compiler_exe="gcc"):
     try:
         _, out = _execute('%s -dumpversion' % compiler_exe)
@@ -159,10 +146,13 @@ def _get_default_compiler(output):
     if cc or cxx:  # Env defined, use them
         output.info("CC and CXX: %s, %s " % (cc or "None", cxx or "None"))
         command = cc or cxx
-        if _is_real_compiler("gcc", command):
-            return _gcc_compiler(output, command)
-        if _is_real_compiler("clang", command.lower()):
-            return _clang_compiler(output, command)
+        if "gcc" in command or "clang" in command.lower():
+            _, out = _execute("%s --version" % command)
+            out = out.lower()
+            if "gcc" in out:
+                return _gcc_compiler(output, command)
+            if "clang" in out:
+                return _clang_compiler(output, command)
         if platform.system() == "SunOS" and command.lower() == "cc":
             return _sun_cc_compiler(output, command)
         # I am not able to find its version

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -25,14 +25,14 @@ def _execute(command):
 
 def _gcc_compiler(output, compiler_exe="gcc"):
 
-    if platform.system() == "Darwin":
-        # In Mac OS X check if gcc is a fronted using apple-clang
-        _, out = _execute("%s --version" % compiler_exe)
-        out = out.lower()
-        if "clang" in out:
-            return None
-
     try:
+        if platform.system() == "Darwin":
+            # In Mac OS X check if gcc is a fronted using apple-clang
+            _, out = _execute("%s --version" % compiler_exe)
+            out = out.lower()
+            if "clang" in out:
+                return None
+
         _, out = _execute('%s -dumpversion' % compiler_exe)
         compiler = "gcc"
         installed_version = re.search("([0-9](\.[0-9])?)", out).group()

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -23,6 +23,19 @@ def _execute(command):
     return proc.returncode, "".join(output_buffer)
 
 
+def _is_real_compiler(compiler_name, compiler_exe):
+    """
+    In some platforms gcc compiler is not a real gcc, i.e: in Mac OS X gcc is clang by default.
+    It can also be a problem if gcc is a symlink to clang or the other way around.
+    """
+
+    try:
+        _, out = _execute("%s --version" % compiler_exe)
+        return compiler_name in out
+    except Exception:
+        return False
+
+
 def _gcc_compiler(output, compiler_exe="gcc"):
     try:
         _, out = _execute('%s -dumpversion' % compiler_exe)
@@ -49,7 +62,7 @@ def _clang_compiler(output, compiler_exe="clang"):
             compiler = "apple-clang"
         elif "clang version" in out:
             compiler = "clang"
-        installed_version = re.search("([0-9]\.[0-9])", out).group()
+        installed_version = re.search("version ([0-9]\.[0-9])", out).group(1)
         if installed_version:
             output.success("Found %s %s" % (compiler, installed_version))
             return compiler, installed_version
@@ -146,9 +159,9 @@ def _get_default_compiler(output):
     if cc or cxx:  # Env defined, use them
         output.info("CC and CXX: %s, %s " % (cc or "None", cxx or "None"))
         command = cc or cxx
-        if "gcc" in command:
+        if _is_real_compiler("gcc", command):
             return _gcc_compiler(output, command)
-        if "clang" in command.lower():
+        if _is_real_compiler("clang", command.lower()):
             return _clang_compiler(output, command)
         if platform.system() == "SunOS" and command.lower() == "cc":
             return _sun_cc_compiler(output, command)

--- a/conans/client/conf/detect.py
+++ b/conans/client/conf/detect.py
@@ -49,7 +49,7 @@ def _clang_compiler(output, compiler_exe="clang"):
             compiler = "apple-clang"
         elif "clang version" in out:
             compiler = "clang"
-        installed_version = re.search("version ([0-9]\.[0-9])", out).group(1)
+        installed_version = re.search("([0-9]\.[0-9])", out).group()
         if installed_version:
             output.success("Found %s %s" % (compiler, installed_version))
             return compiler, installed_version
@@ -146,13 +146,19 @@ def _get_default_compiler(output):
     if cc or cxx:  # Env defined, use them
         output.info("CC and CXX: %s, %s " % (cc or "None", cxx or "None"))
         command = cc or cxx
-        if "gcc" in command or "clang" in command.lower():
+        if "gcc" in command:
+            if platform.system() != "Darwin":
+                return _gcc_compiler(output, command)
+
+            # In Mac OS X check if gcc is a fronted using apple-clang
             _, out = _execute("%s --version" % command)
             out = out.lower()
-            if "gcc" in out:
+            if "clang" not in out:
                 return _gcc_compiler(output, command)
-            if "clang" in out:
-                return _clang_compiler(output, command)
+            output.warn("%s detected as a frontend using apple-clang, skipping it to use native apple-clang" % command)
+            command = "clang"  # Force use of clang when gcc is detected as a fronted of apple-clang
+        if "clang" in command.lower():
+            return _clang_compiler(output, command)
         if platform.system() == "SunOS" and command.lower() == "cc":
             return _sun_cc_compiler(output, command)
         # I am not able to find its version

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -53,3 +53,5 @@ class DetectTest(unittest.TestCase):
         # result is a list of tuples (name, value) so converting it to dict
         result = dict(result)
         self.assertEquals(result.get("compiler", None), "apple-clang")
+        self.assertIn("gcc detected as a frontend using apple-clang", output)
+        self.assertIsNotNone(output.warn)

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -8,18 +8,17 @@ from conans.test.utils.tools import TestBufferConanOutput
 
 class DetectTest(unittest.TestCase):
 
-    platform_default_compilers = {
-        "Linux": "gcc",
-        "Darwin": "apple-clang",
-        "Windows": "Visual Studio"
-    }
-
     def detect_default_compilers_test(self):
+        platform_default_compilers = {
+            "Linux": "gcc",
+            "Darwin": "apple-clang",
+            "Windows": "Visual Studio"
+        }
         output = TestBufferConanOutput()
         result = detect_defaults_settings(output)
         # result is a list of tuples (name, value) so converting it to dict
         result = dict(result)
-        platform_compiler = self.platform_default_compilers.get(platform.system(), None)
+        platform_compiler = platform_default_compilers.get(platform.system(), None)
         if platform_compiler is not None:
             self.assertEquals(result.get("compiler", None), platform_compiler)
 
@@ -31,6 +30,4 @@ class DetectTest(unittest.TestCase):
             result = detect_defaults_settings(output)
         # result is a list of tuples (name, value) so converting it to dict
         result = dict(result)
-        platform_compiler = self.platform_default_compilers.get(platform.system(), None)
-        self.assertIsNotNone(platform_compiler)
-        self.assertEquals(result.get("compiler", None), platform_compiler)
+        self.assertEquals(result.get("compiler", None), "apple-clang")

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -39,7 +39,8 @@ class DetectTest(unittest.TestCase):
 
         if "clang" not in output:
             # Not test scenario gcc should display clang in output
-            return
+            # see: https://stackoverflow.com/questions/19535422/os-x-10-9-gcc-links-to-clang
+            raise Exception("Apple gcc doesn't point to clang with gcc frontend anymore! please check")
 
         output = TestBufferConanOutput()
         with tools.environment_append({"CC": "gcc"}):

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -1,6 +1,6 @@
 import platform
+import subprocess
 import unittest
-from subprocess import Popen, PIPE, STDOUT
 
 from conans import tools
 from conans.client.conf.detect import detect_defaults_settings
@@ -31,23 +31,14 @@ class DetectTest(unittest.TestCase):
         if platform.system() != "Darwin":
             return
 
-        def _execute(command):
-            proc = Popen(command, shell=True, bufsize=1, stdout=PIPE, stderr=STDOUT)
+        try:
+            output = subprocess.check_output(["gcc", "--version"], stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError:
+            # gcc is not installed or there is any error (no test scenario)
+            return
 
-            output_buffer = []
-            while True:
-                line = proc.stdout.readline()
-                if not line:
-                    break
-                # output.write(line)
-                output_buffer.append(str(line))
-
-            proc.communicate()
-            return proc.returncode, "".join(output_buffer)
-
-        proc_return, output = _execute("gcc --version")
-        if proc_return != 0 or "clang" not in output:
-            # Not test scenario (gcc should display clang in output
+        if "clang" not in output:
+            # Not test scenario gcc should display clang in output
             return
 
         output = TestBufferConanOutput()

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -37,7 +37,7 @@ class DetectTest(unittest.TestCase):
             # gcc is not installed or there is any error (no test scenario)
             return
 
-        if "clang" not in output:
+        if b"clang" not in output:
             # Not test scenario gcc should display clang in output
             # see: https://stackoverflow.com/questions/19535422/os-x-10-9-gcc-links-to-clang
             raise Exception("Apple gcc doesn't point to clang with gcc frontend anymore! please check")

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -1,15 +1,36 @@
-import unittest
-from conans.test.utils.tools import TestBufferConanOutput
-from conans.client.conf.detect import detect_defaults_settings
 import platform
+import unittest
+
+from conans import tools
+from conans.client.conf.detect import detect_defaults_settings
+from conans.test.utils.tools import TestBufferConanOutput
 
 
 class DetectTest(unittest.TestCase):
 
-    def detect_test(self):
+    platform_default_compilers = {
+        "Linux": "gcc",
+        "Darwin": "apple-clang",
+        "Windows": "Visual Studio"
+    }
+
+    def detect_default_compilers_test(self):
         output = TestBufferConanOutput()
         result = detect_defaults_settings(output)
-        if platform.system() == "Linux":
-            for (name, value) in result:
-                if name == "compiler":
-                    self.assertEquals(value, "gcc")
+        # result is a list of tuples (name, value) so converting it to dict
+        result = dict(result)
+        platform_compiler = self.platform_default_compilers.get(platform.system(), None)
+        if platform_compiler is not None:
+            self.assertEquals(result.get("compiler", None), platform_compiler)
+
+    @unittest.skipIf(platform.system() != "Darwin", "Test only running in Mac OS X")
+    def detect_default_in_mac_os_using_gcc_as_default_test(self):
+        # See: https://github.com/conan-io/conan/issues/2231
+        output = TestBufferConanOutput()
+        with tools.environment_append({"CC": "gcc"}):
+            result = detect_defaults_settings(output)
+        # result is a list of tuples (name, value) so converting it to dict
+        result = dict(result)
+        platform_compiler = self.platform_default_compilers.get(platform.system(), None)
+        self.assertIsNotNone(platform_compiler)
+        self.assertEquals(result.get("compiler", None), platform_compiler)

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -25,9 +25,12 @@ class DetectTest(unittest.TestCase):
 
     def detect_default_in_mac_os_using_gcc_as_default_test(self):
         """
-        In some Mac OS X gcc is in reality clang.
+        Test if gcc in Mac OS X is using apple-clang as frontend
         """
         # See: https://github.com/conan-io/conan/issues/2231
+        if platform.system() != "Darwin":
+            return
+
         def _execute(command):
             proc = Popen(command, shell=True, bufsize=1, stdout=PIPE, stderr=STDOUT)
 
@@ -44,7 +47,7 @@ class DetectTest(unittest.TestCase):
 
         proc_return, output = _execute("gcc --version")
         if proc_return != 0 or "clang" not in output:
-            # Not test scenario (gcc should display that it's clang)
+            # Not test scenario (gcc should display clang in output
             return
 
         output = TestBufferConanOutput()

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -47,6 +47,7 @@ class DetectTest(unittest.TestCase):
             result = detect_defaults_settings(output)
         # result is a list of tuples (name, value) so converting it to dict
         result = dict(result)
-        self.assertEquals(result.get("compiler", None), "apple-clang")
+        # No compiler should be detected
+        self.assertIsNone(result.get("compiler", None))
         self.assertIn("gcc detected as a frontend using apple-clang", output)
-        self.assertIsNotNone(output.warn)
+        self.assertIsNotNone(output.error)

--- a/conans/test/util/detect_test.py
+++ b/conans/test/util/detect_test.py
@@ -1,5 +1,6 @@
 import platform
 import unittest
+from subprocess import Popen, PIPE, STDOUT
 
 from conans import tools
 from conans.client.conf.detect import detect_defaults_settings
@@ -22,9 +23,30 @@ class DetectTest(unittest.TestCase):
         if platform_compiler is not None:
             self.assertEquals(result.get("compiler", None), platform_compiler)
 
-    @unittest.skipIf(platform.system() != "Darwin", "Test only running in Mac OS X")
     def detect_default_in_mac_os_using_gcc_as_default_test(self):
+        """
+        In some Mac OS X gcc is in reality clang.
+        """
         # See: https://github.com/conan-io/conan/issues/2231
+        def _execute(command):
+            proc = Popen(command, shell=True, bufsize=1, stdout=PIPE, stderr=STDOUT)
+
+            output_buffer = []
+            while True:
+                line = proc.stdout.readline()
+                if not line:
+                    break
+                # output.write(line)
+                output_buffer.append(str(line))
+
+            proc.communicate()
+            return proc.returncode, "".join(output_buffer)
+
+        proc_return, output = _execute("gcc --version")
+        if proc_return != 0 or "clang" not in output:
+            # Not test scenario (gcc should display that it's clang)
+            return
+
         output = TestBufferConanOutput()
         with tools.environment_append({"CC": "gcc"}):
             result = detect_defaults_settings(output)

